MVP - Pulling Images onto disk

daniel5151 opened the pull request:
Tentative milestone: be able to specify an image (in docker image reference format), and download it's manifest + layers onto disk.

As it turns out, this is quite the yak-shave, requiring a whole bunch of things to execute correctly.

This PR will gradually get updated as more bits of functionality are implemented.

daniel5151 committed:
initial commit - implement reference parsing

daniel5151 committed:
oci_image: initial commit - a few basic types

daniel5151 committed:
oci_distribution: initial commit - a few basic types

daniel5151 committed:
basic client + auth skeleton

Just enough functionality to authenticate with the public docker
registry, and pull an image manifest.

Error handling is _bad_, and needs to be redone properly. There are way
too many panics and unwraps (but to be fair, this is a MVP for proving
out flows!)

daniel5151 committed:
oci_image: add remaining types

still a couple of TODOs left, mainly related to using proper types for
things like Digests and Time values.

daniel5151 committed:
robust www_authenticate header parsing

gone with the unreadable regex! now we have a proper, (mostly) standards
conforming grammar!

There are a couple more edge cases to iron out, but honestly, the effort
required to fix them is possibly non-trivial, and it's already robust
_enough_

myagley commented on containrs/Cargo.toml:
```diff
@@ -0,0 +1,7 @@
+    "containrs",
+    "containrs-cli",
+    "oci_distribution",
+    "oci_image",
```
huge nit, but do you want to make the `oci` crates also kebab-case instead of snake_case to stay consistent?

myagley commented on containrs/containrs-cli/src/main.rs:
```diff
@@ -0,0 +1,120 @@
+    let app_m = App::new("containrs-cli")
+        .setting(AppSettings::ArgRequiredElseHelp)
+        .version("0.1.0")
+        .author("Daniel Prilik <danielprilik@gmail.com>")
```
I think for author we usually use the royal "Azure IoT Edge Devs".

daniel5151 replied:
Oops, I knew that, just an oversight from copying-pasting autogenerated cargo.tomls.
I'll update the rest of them as well.

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -0,0 +1,108 @@
+    /// Wrapper around [HyperClient::request] which authenticates outbound
+    /// requests
+    pub async fn request(&mut self, mut req: Request<Body>) -> Result<Response<Body>> {
+        let headers = match self.store.get(req.uri()) {
```
Is there something that puts into this HashMap or is that still coming?

daniel5151 replied:
Yep, that's an oversight on my part.

As a sidenote, I'm fairly sure this "cache-by-URI" scheme isn't the right way to do it, and instead, a more "intrusive" inspection of the URI might be required (i.e: validate by endpoint, not endpoint + concrete arguments).

I'll be tweaking this as I learn more about how the API functions "in the real world"

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -0,0 +1,108 @@
+
+    /// Wrapper around [HyperClient::get] which authenticates outbound requests
+    pub async fn get(&mut self, uri: Uri) -> Result<Response<Body>> {
+        self.request(Request::get(uri).body(Body::default())?).await
```
This is kind of a nit but you can probably just return this future instead of making the one liner async/await. This would normally get flagged in other async/await languages. Should be straightforward with `impl trait`

daniel5151 replied:
One small pro by leaving it as-is would be that it matches the underlying signature of HyperClient::get exactly, which communicates intent a bit more.

That said, I'm cool with changing it if you think that's not too important.

myagley replied:
We don't need to update it.

myagley commented on containrs/containrs/src/auth/www_authenticate.pest:
```diff
@@ -0,0 +1,87 @@
+// # from RFC-2617 (HTTP Basic and Digest authentication)
```
You are getting good at PEG grammars! :)

myagley commented on containrs/containrs/src/auth/www_authenticate.rs:
```diff
@@ -0,0 +1,495 @@
+}
+
+#[cfg(test)]
+mod tests {
```
This (and the reference parser) would be a perfect candidate to use proptest on. We don't need to focus on it now but something to circle back on later. Proptest would give a lot of confidence about the correctness.

daniel5151 replied:
Oh yeah, for sure! I'd love to iron out issues with those sub-crates, since it would _suck_ trying to diagnose an issue stemming from mis-parsed text.

I'll add a // TODO block in both of the test-blocks as a reminder

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -0,0 +1,140 @@
+    domain: String,
+}
+
+impl<C: Connect + 'static> Client<C> {
```
The `'static` is required for the futures executor?

daniel5151 replied:
I'm not sure what it's required for _exactly_, but the compiler gets mad if it's not there.
Moreover, if you grep for `Connect + 'static` in edgelet, you'll see it's a fairly common bound used in similar client-wrappers

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -0,0 +1,140 @@
+    /// Check that basic authentication works (by pinging the base URL:
+    /// <domain>/v2/)
+    pub async fn check_basic_auth(&mut self) -> Result<bool> {
+        self.check_authentication("/v2/", Method::GET).await
```
Same comment about the one liner and async/await

daniel5151 replied:
Hmm, this could just be my lack of experience, but I think the async/await version is more readable than

```rust
pub fn check_basic_auth(&mut self) -> impl Future<Output = Result<bool>> + '_ {
    self.check_authentication("/v2/", Method::GET)
}
```

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -0,0 +1,140 @@
+        paginate: Option<Paginate>,
+    ) -> Result<Option<(Catalog, Option<Paginate>)>> {
+        if let Some(paginate) = paginate {
+            let _ = paginate;
```
[nit]  You can probably do the following to silence the compiler:
```rust
if let Some(_paginate) = paginate {
    unimplemented!("implement _catalog pagination")
}
```

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -0,0 +1,140 @@
+    pub async fn get_manifest(
+        &mut self,
+        name: &str,
+        reference: &ReferenceKind,
```
Should this take a full canonicalized reference? Then you know every thing is valid.

```rust
pub async fn get_manifest(&mut self, reference: &Reference) -> Result<Manifest>
```

daniel5151 replied:
...oops. Yeah, absolutely. Otherwise, what's the point of having the type in the first place lol

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -0,0 +1,140 @@
+    }
+}
+
+use serde::{Deserialize, Serialize};
```
[nit] move this to the top?

myagley commented on containrs/containrs/src/util/hyper.rs:
```diff
@@ -0,0 +1,33 @@
+use crate::Result;
+
+/// Utility methods for working with [hyper::Body]
+#[async_trait]
```
Interesting. What is this doing? Are traits not able to have `async` methods?

daniel5151 replied:
Yep, async traits aren't a thing yet :disappointed: 

Thankfully, dtolnay wrote [async_trait](https://crates.io/crates/async-trait), which implements the feature. It's some real wizardry, but the end result is incredibly seamless.

myagley commented on containrs/containrs/src/reference/mod.rs:
```diff
@@ -0,0 +1,527 @@
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum ReferenceKind {
+    Tag(String),
+    Digest(String), // TODO: replace with proper digest type
```
Are these mutually exclusive? It seems like they should be because having a digest kind of negates the usefulness of a tag

myagley replied:
I guess the question is should the `FromStr` and `Display` impls for the Reference type be reversible/reproducible. Should you get back from `Display` exactly what you parsed?

daniel5151 replied:
The grammar lets you specify both at the same time, but when communicating with the v2 API, you pass either a digest or a tag, giving the digest priority if both are specified.

As such, I _personally_ think that it's fine to "lose" the Tag information if both are specified, as it is the less-specific of the two kinds.

myagley commented on containrs/oci_image/src/v1/config.rs:
```diff
@@ -0,0 +1,142 @@
+pub struct RootFS {
+    /// Type is the type of the rootfs.
+    #[serde(rename = "type")]
+    pub rootfs_type: String, // cannot use `type` as name. it's a reserved word
```
You can with raw identifiers I believe, but it's up to you if that's too ugly.

```rust
pub r#type: String,
```

daniel5151 replied:
yeaaaaah, I knowwww, but if I saw a raw identifier out in the wild, I wouldn't be too happy.

myagley replied:
Yeah, they are ugly

myagley commented on containrs/oci_image/src/v1/config.rs:
```diff
@@ -0,0 +1,142 @@
+
+    /// RootFS references the layer content addresses used by the image.
+    #[serde(rename = "rootfs")]
+    pub root_fs: RootFS,
```
[huge nit] I'd probably just keep the `rootfs` for the rust variable as well.

daniel5151 replied:
Nah, that's totally fair. I just blindly accepted a bunch of clippy lints when converting between the CamelCase Go types, and assumed it would all work out.

daniel5151 committed:
kebab-case package names

daniel5151 committed:
addressing comments

daniel5151 committed:
extract www-authenticate and docker-reference as separate crates

daniel5151 committed:
major error handling overhaul

i.e: actually handle errors now!

daniel5151 commented:
@myagley 
Just pushed my first-crack at error handling. It's been a real journey, but I think it's pretty okay as it stands. Give it a look through and tell me if anything looks funny.

P.S: give me a :+1: on any conversations above that I can mark as resolved.

myagley commented on containrs/containrs/src/auth/docker.rs:
```diff
@@ -53,26 +56,42 @@ pub async fn auth_flow<C: Connect + 'static>(
+    let res = client
+        .get(uri)
+        .await
+        .context(AuthError::AuthServerNoResponse)?;
```
I know the postfix syntax for await was controversial but its pretty nice to do things like this (instead of having to add parens)

daniel5151 replied:
yeah, I had my reservations when I heard about it, but after working with it for the past little while, I really like it

myagley commented on containrs/containrs/src/auth/error.rs:
```diff
@@ -0,0 +1,51 @@
+
+    #[fail(display = "Auth server token is invalid")]
+    AuthServerInvalidToken,
+}
```
looks good to me

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -1,59 +1,97 @@
+use failure::{Fail, ResultExt};
```
As a resource, the webmachine state machine is pretty good for understanding the possible flows of HTTP: https://raw.githubusercontent.com/wiki/webmachine/webmachine/images/http-headers-status-v3.png

This flow diagram helped me really understand HTTP.

daniel5151 replied:
Oh wow, that seems complicated, but incredibly useful. I'll definitely bookmark that and refer to it

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -64,77 +102,88 @@ impl<C: Connect + 'static> Client<C> {
+                | StatusCode::FORBIDDEN => Err(new_api_error(status, res).await),
+                _ => {
+                    res.dump_to_debug().await;
+                    Err(ErrorKind::ApiUnexpectedStatus(status).into())
```
This is a nit, but Forbidden is probably going to be a relatively common error condition for these calls. We may want to tweak the error messages for the common case.

daniel5151 replied:
Oh yeah, I don't doubt that.

I'll revisit these error messages once I get Oauth2 working, and I can communicate with Azure Container Registries as well. That way, I'll have two different implementations of the API to ping, and that should make it more clear how the error messages should look.

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -64,77 +102,88 @@ impl<C: Connect + 'static> Client<C> {
+/// value, returns a Error(ErrorKind::ApiError) who's context is either the
+/// ApiErrors JSON itself, or, if the JSON was malformed, a
+/// ErrorKind::ApiMalformedJSON.
+async fn new_api_error(status: StatusCode, res: Response<Body>) -> Error {
```
You can probably get the status from the `Response<Body>`? Could this signature be something like:

```rust
async fn new_api_error(res: Response<Body>) -> Error {
```

Because of the async nature this is impossible but this looks like a `impl From<Response<Body>> for Error`

daniel5151 replied:
Oops, you're absolutely right. 
Minor oversight on my part.

And you're totally right, this is essentially just a workaround for not being able to do `impl From<Response<Body>> for Error`

myagley commented on containrs/containrs/src/error.rs:
```diff
@@ -0,0 +1,109 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self.0)
+    }
+}
```
looks good to me

myagley commented on containrs/www-authenticate/src/error.rs:
```diff
@@ -0,0 +1,19 @@
+#[derive(Debug, Fail)]
+pub enum WWWAuthenticateError {
+    #[fail(display = "Failed to parse WWW-Authenticate Header: {}", _0)]
+    Pest(PestError<Rule>),
```
We may want to call this something like `Parse` instead of `Pest`?

daniel5151 committed:
address minor comments

daniel5151 committed:
add transport-scheme parameter to client

this fixes errors when a user includes a custom registry in the docker
reference string

daniel5151 committed:
add username/password based OAuth2 auth

This enables authenticating with Azure Container Registries via the
Admin account.

daniel5151 committed:
add /tags endpoint, pagination, and cli refactor

I've split most client methods into 2 variants, a get_raw_X, and an
associated get_X. get_raw_X handles the nitty-gritty details of
performing the request, and getting the raw data. get_X is a wrapper
around get_raw_X which also performs some additional validation on the
data (e.g: ensuring it is valid JSON).

In turn, I've refactored the CLI to group the get_raw_X methods under
the "raw" subcommand. These will dump raw data directly to stdout.

High-level comamnds, such as download, will remain at the top-level of
the CLI command hierarchy.

daniel5151 committed:
remove test-auth CLI argument

(it's served it's purpose)

daniel5151 committed:
minor error handling tweak

daniel5151 committed:
fix tests

daniel5151 committed:
first pass at blob pulling

Not having Hyper automatically follow redirects is... annoying. There's
a surprisingly complex flow that I would have to implement if I wanted a
robust redirect implementation.

I'm seriously debating switching HTTP clients to something a bit
higher-level, such as reqwest or surf. The tricky bit is that I'm not
sure which to pick, as each has it's owns pros and cons:

request:
- Pro: reqwest is well-established, and known good. Plus, it has a
robust redirect policy built-in!
- Con: it looks like the next release will have some major refactoring
and breaking changes (while adding a bunch of new features), so if I use
reqwest right now, upgrading will be annoying.
- Con: no support for middleware

surf:
- Pro: it's the new crate on the block, and seems to have a
significantly cleaner interface.
- Pro: it has a robust middleware system, which would make wrapping
requests with authentication headers reall clean
- Con: it's new, and not battle tested. While some projects have started
using it, I'm not sure if it will be maintained actively (the last major
batch of commits was >1mo ago, whereas reqwest is being actively
developed)
- Con: no built-in redirect policy (though there is an open PR
implementing a redirect middleware)

daniel5151 committed:
minor error tweak

daniel5151 committed:
first pass at image download flow

I mean, it works, but not really as intended.
The client API needs to be rewritten somewhat to enable multiple
simultaneous downloads. Since every get_X call takes the client as &mut
self, there isn't a clean way (that I know of) to get multiple requests
going at once.

daniel5151 committed:
switch to beta toolchain

daniel5151 committed:
tweak client API to enable simultaneous downloads

It's still not _as_ parallel as I'd like (requiring synchronously
kicking off downloading each blob), but it's a great improvement.

daniel5151 committed:
removed file I/O from containrs core

instead, the download_image flow returns a collection of Body streams,
which the client can use however they like

daniel5151 committed:
add progress indicators for donwloads

get_blob returns the total size of the body as well now.

daniel5151 committed:
minor progress tweak for raw blob cmd

daniel5151 committed:
port to reqwest + client async/await improvements

Porting things over to reqest took a while, but it's definately worth it.
Redirect handling is really hard.

I've also reworked the way authentication caching works, switching
things over to a scope-based system, which results in a _lot_ more cache
hits.

While this is a great improvement, there are still a couple of hot-spots
in the code that could be improved:
- if two requests with disjoint scopes are performed, and each requires
a authentication handshake, one of them will block while the other
completes it's auth handshake. see the TODO in auth/mod.rs:60
- there's probably a way to return the raw Result<Blob> futures from the
dowload_image flow helper, but my Futures-fu isn't good enough to get
the types/lifetimes to work right.
- Right now, the scope header in the www-authenticate header is totally
ignored, and the scope cache uses the "best guess" scope given by the
various client API methods. This works okay, but it would be better to
actually parse that scope header and use it to sanity-check the expected
scopes. Also, if the server returns multiple scopes, we can pre-populate
the cache with those additional entries.

daniel5151 committed:
improve scope-caching infrastructure

oh look, another parser. how fun!
now that scopes aren't totally opaque strings, it opens the door to much
improved caching, specifically, in cases where the server resuts scopes
with more than one resource / action.

daniel5151 committed:
rework scope-based auth header caching

ACR broke my invariant, and forced me to switch up how I cache scopes.

myagley commented on containrs/containrs-cli/src/main.rs:
```diff
@@ -16,103 +29,400 @@ async fn main() {
         .arg(
             Arg::with_name("default-registry")
-                .help("Default registry (defaults to https://registry-1.docker.io/v2/)")
+                .help("Default registry (defaults to \"registry-1.docker.io\")")
```
I think there's a way to specify a default with clap

daniel5151 replied:
There probably is, but I eventually want to read config data from disk, so I'm leaving it separate for now

myagley commented on containrs/containrs-cli/src/main.rs:
```diff
@@ -16,103 +29,400 @@ async fn main() {
-                .about("Retrieve a sorted, json list of repositories available in the registry"),
+            SubCommand::with_name("raw")
+                .setting(AppSettings::SubcommandRequiredElseHelp)
+                .about("Retrieve raw responses from various endpoints")
```
Nice!

myagley replied:
This will be super helpful for debugging

myagley commented on containrs/containrs-cli/src/main.rs:
```diff
@@ -16,103 +29,400 @@ async fn main() {
-    let docker_compat = true;
+        .unwrap_or("registry-1.docker.io");
+    // TODO: this should probably be a more robust check
+    let docker_compat = default_registry.contains("docker");
```
What does docker compatibility change?

daniel5151 replied:
it's passed to the docker-reference canonicalize function, and mainly deals with some hacky docker-isms that other registries don't need to deal with (e.g: the raw repo name `ubuntu` actually corresponds to the `library/ubuntu` repo on docker, so the compat flag will prepend `library` to repos that consist of a single path component)

myagley commented on containrs/containrs-cli/src/main.rs:
```diff
@@ -16,103 +29,400 @@ async fn main() {
-                None => println!("Registry doesn't support the _catalog endpoint"),
+                ("tags", Some(sub_m)) => {
+                    // won't panic, as this is a required argument
+                    let repo = sub_m.value_of("repo").unwrap();
```
You might want to use an `expect` instead of the comment and `unwrap`. Then it's guaranteed to follow the code around.

myagley commented on containrs/containrs-cli/src/main.rs:
```diff
@@ -16,103 +29,400 @@ async fn main() {
+            // asyncronously dump the bodies to disk
+
+            lazy_static! {
+                static ref MEDIA_TYPE_TO_FILE_EXT: HashMap<&'static str, &'static str> = {
```
Can we move this static info to the top of the file?

myagley commented on containrs/containrs/Cargo.toml:
```diff
@@ -5,16 +5,21 @@ authors = ["Azure IoT Edge Devs"]
 serde_urlencoded = "0.5"
+tokio = "0.2.0-alpha.5"
+
+reqwest = { git = "https://github.com/seanmonstar/reqwest.git", features = ["unstable-stream", "json"] }
```
We may want to pin this to a particular commit so that it doesn't sporadically break the build? Or does the `Cargo.lock` have the sha?

daniel5151 replied:
Cargo.lock takes care of that :+1:

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -1,145 +1,228 @@
             client,
             creds,
-            store: HashMap::new(),
+            scope_cache: RwLock::new(ScopeCache::new()),
```
To verify my understanding, the point of the scope cache is limit the number of auth flows that need to be done?

daniel5151 replied:
Yeah. For example, after checking the logs, I realized that the exact same scope is used across all requests when pulling an image manifest + subsequent blobs. Caching based on scope is nice, since only a single auth flow ends up being required per download flow.

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -1,145 +1,228 @@
+        if let Some(headers) = self
+            .scope_cache
+            .read()
+            .map_err(|_| AuthError::CacheLock)?
```
I think its fairly common to panic on a lock error because this only returns an error if the lock is poisoned?

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -1,145 +1,228 @@
+            .map_err(|_| AuthError::CacheLock)?
+            .get(expected_scope)
+        {
+            // TODO?: store expiration time alongside scope, and check for expiry
```
Yes, we should probably do that :)

myagley replied:
And make the cache handle the expiry.

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -1,145 +1,228 @@
+        // do one more check to be _certain_ that no other requests have already
+        // performed the authentication flow in the time between the first check and
+        // acquiring the write lock
+        if let Some(headers) = scope_cache.get(expected_scope) {
```
:+1:

myagley commented on containrs/containrs/src/auth/scopecache.rs:
```diff
@@ -0,0 +1,32 @@
+    /// Insert a new scope-headers pair into the cache.
+    pub fn insert(&mut self, scope: Scope, headers: HeaderMap) {
+        self.map.push((scope, headers))
+    }
```
Does the order of the `Vec` matter? If you are trying to find a scope that is a superset is it ok to find the first one or does it need to be the broadest scope?

daniel5151 replied:
Hmmm, as it stands, I don't think order matters, but honestly, I'm not entirely sure. This implementation of the cache really isn't great, and could stand to be improved significantly. The tricky bit is that I won't know the best way to structure the cache until I get a better idea of what a typical use case for the library looks like, and that's something that only time and more testing can provide.

myagley replied:
Sure makes sense.

myagley commented on containrs/containrs/src/flows/download_image.rs:
```diff
@@ -0,0 +1,58 @@
+    pub manifest_json: Blob,
+    pub config_json: Blob,
+    pub layers: Vec<(Blob, ociv1::Descriptor)>,
+}
```
Is this a public type? We should probably use getters for it instead of exposing the struct fields directly?

daniel5151 replied:
This touches on a slightly larger question: should "flows" be a part of the core library?

If you look at what exactly the download_images function is doing, you'll see that it's not using any library internals, and is actually just orchestrating some actions on a client. As it stands, this function could be copied and pasted into user-code with no issue.

I'm starting to think that's actually the way to go, since it leaves the door to more fine-grained control over the download process (i.e: don't download layers we already have the SHA of) .

... yeah, after writing this out, I think i'm actually gonna remove flows from the core library, and implement this method in the CLI

myagley commented on containrs/rust-toolchain:
```diff
@@ -1 +1 @@
-nightly
\ No newline at end of file
+beta
```
awesome!

daniel5151 committed:
address comments + move flows out of lib

see https://github.com/daniel5151/iotedge/pull/1#discussion_r331149332
for rationale on the flows removal

daniel5151 committed:
create oci-digest

finally, we can remove all the TODOs related to not using a proper
digest type!

daniel5151 committed:
replace String with oci-digest + blob validation!

daniel5151 committed:
polish up blob validation

daniel5151 committed:
move oci-digest closer to spec compliance

daniel5151 committed:
oci-image improvements

lots of additional inline comments, and additional validation checks.

daniel5151 committed:
implement oci image layout spec

i.e: by running `tar cvf <image>.tar *` in the download outdir, the
resulting tar archive can be imported into Docker / containerd!

daniel5151 committed:
remove tools folder

it's served it's purpose

daniel5151 committed:
add misc utlity methods and derives

daniel5151 committed:
overhaul concurrency in client + authclient

namely, switching to Futures aware synchronization primitives (i.e:
futures::lock::Mutex), and implementing a more efficient concurrent
hashmap (with both inner and outer locking).

The concurrency is kinda tricky to reason about, but I _think_ the
current implementation is correct?

I bet you could get rid of some of those `Arc`s I have scattered around
though...

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -90,38 +92,45 @@ impl AuthClient {
+        // preventing any other tasks from calling .get() in the brief period of
+        // time between insertion and lock acquisition.
+        scope_cache.insert(expected_scope.clone(), HeaderMap::new());
+        let headers = scope_cache.get(expected_scope).unwrap();
```
We should handle this error or use `expect`.

daniel5151 replied:
Sure, I can switch it to an expect, though I can't see a situation where it could fail

daniel5151 replied:
in fact, I'm not even sure what to put in the expect message :sweat_smile:

myagley replied:
That you don't expect it to fail? :)

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -90,38 +92,45 @@ impl AuthClient {
+        scope_cache.insert(expected_scope.clone(), HeaderMap::new());
+        let headers = scope_cache.get(expected_scope).unwrap();
+        let mut headers = headers.lock().await;
+        std::mem::drop(scope_cache);
```
Could this deadlock? Is the only other place this lock is held is in the if statement above, which drops the lock before taking the headers lock?

daniel5151 replied:
The lock is acquired _before_ the if-statement (on line 101), so I'm pretty sure this is sound?

myagley replied:
Yeah, I don't think its unsound, just wanted to check my understanding.

myagley commented on containrs/containrs/src/auth/mod.rs:
```diff
@@ -138,91 +147,163 @@ impl AuthClient {
+    pub fn get(&self, scope: &Scope) -> Option<Arc<Mutex<HeaderMap>>> {
+        for (s, headers) in self.map.iter() {
+            if s.is_superset(scope) {
+                return Some(Arc::clone(headers));
```
I may have asked this before, but order of this map doesn't matter?

daniel5151 replied:
You have, and no it doesn't.
That said, this lookup code is _bad_. I'm sure there's a more efficient way to structure it, through it's a bit tricky, given that it's not a simple key-value kind of cache, and is based on scopes (which are sets).

myagley commented on containrs/containrs/src/client.rs:
```diff
@@ -230,19 +232,21 @@ impl Client {
                 }
 
-                if !self.supports_range_header.borrow().unwrap() {
+                if self.supports_range_header.load(Ordering::Relaxed) == 0 {
```
I don't know enough about atomics to really review this, but does it make sense that all accesses use the same Ordering? Also, I'm assuming there's no atomic bool?

daniel5151 replied:
I've taken a look at this code with fresh eyes, and no, it's totally bogus :grimacing: 
I'm not sure what I was thinking here tbh. If it's supposed to prevent multiple threads from performing a head request, then it totally fails. Yikes.

I'm gonna rewrite this asap.

There is an atomic bool, but I need a tri-state variable (true/false if the server supports ranges, and an additional "unknown" state which designates that the server still needs to be queried)

daniel5151 committed:
add test command to cli for profiling scope cache

daniel5151 committed:
address comments + fix bogus range header locking

daniel5151 committed:
re-export useful mods from containrs core

daniel5151 committed:
support blob downloads via descriptor URLs

ran into this edge case when downloading the windows server issue.

The implementation got a bit hairy, no thanks to the silly fact that
Ranges can't be cloned! Oof.

daniel5151 committed:
move some mod.rs files up a dir

daniel5151 committed:
port oci-runtime types from Go to Rust

daniel5151 committed:
reorg and clean up oci-runtime types

daniel5151 committed:
clean up oci-image custom ser/de handling

daniel5151 committed:
move image_layout under v1::util

this is laying the groundwork for additional utilities, such as the OCI
image config to OCI runtime config converter.

daniel5151 committed:
prune more-harm-than-good derive(Default) impls

daniel5151 committed:
create oci-common crate

daniel5151 committed:
refactor and cleanup oci-image and oci-runtime

daniel5151 committed:
initial oci image to runtime config conversion

untested, and may have bugs / require tweaking

daniel5151 committed:
async/await hit stable!

Option::flatten() did _not_ hit stable, so had to fix that real quick.

daniel5151 committed:
use std::error::Error in OCI crates + update deps

daniel5151 committed:
docker-reference overhaul + misc bugfixes

as time went on, and I became more familiar with the container
ecosystem, i've realized that 99% of the time, things expect image
references to be docker compatible / canonicalized using standard docker
rules. As such, I've reworked the docker-reference crate to canonicalize
everything by default. This greatly simplifies a whole bunch of code,
and nullifies any tricky configuration requirements surrounding image
reference canonicalization.
